In [1]:
from sklearn import svm, datasets
iris= datasets.load_iris()

In [3]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [6]:
import pandas as pd
df= pd.DataFrame(iris.data, columns= iris.feature_names)
df['flower']= iris.target
df['flower']= df['flower'].apply(lambda x: iris.target_names[x])
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

        flower  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

### **Approach 1: Use train_test_split and manually tune parameters by trial and error**

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(iris.data, iris.target, test_size= 0.3)

In [14]:
model= svm.SVC(kernel= 'rbf', C=30, gamma= 'auto')
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.9777777777777777

### **Approach 2: Use GridSearchCV**

In [15]:
from sklearn.model_selection import GridSearchCV
clf= GridSearchCV(svm.SVC(gamma='auto'),
                  {
                      'C': [1, 30, 50],
                      'kernel': ['rbf', 'linear']
                  },
                  cv= 5, return_train_score=False)
clf.fit(iris.data, iris.target)
clf.cv_results_

{'mean_fit_time': array([0.00587826, 0.00395207, 0.00437989, 0.00131869, 0.00305638,
        0.00132618]),
 'std_fit_time': array([5.87966382e-03, 5.32224692e-03, 3.43084231e-03, 7.73289629e-05,
        2.67208489e-03, 1.26128611e-05]),
 'mean_score_time': array([0.00109286, 0.00225573, 0.0011272 , 0.00245943, 0.00206037,
        0.00105553]),
 'std_score_time': array([6.82437773e-05, 2.60700322e-03, 7.43776402e-05, 2.91208549e-03,
        1.31818672e-03, 6.01677805e-05]),
 'param_C': masked_array(data=[1, 1, 30, 30, 50, 50],
              mask=[False, False, False, False, False, False],
        fill_value=999999),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value=np.str_('?'),
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 30, 'kernel': 'rbf'},
  {'C': 30, 'kernel': 'linear'},
  {'C': 50, 'kernel': 'rbf'},
 

In [16]:
df= pd.DataFrame(clf.cv_results_)
df

mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_C  \
0       0.005878      0.005880         0.001093        0.000068        1   
1       0.003952      0.005322         0.002256        0.002607        1   
2       0.004380      0.003431         0.001127        0.000074       30   
3       0.001319      0.000077         0.002459        0.002912       30   
4       0.003056      0.002672         0.002060        0.001318       50   
5       0.001326      0.000013         0.001056        0.000060       50   

  param_kernel                         params  split0_test_score  \
0          rbf      {'C': 1, 'kernel': 'rbf'}           0.966667   
1       linear   {'C': 1, 'kernel': 'linear'}           0.966667   
2          rbf     {'C': 30, 'kernel': 'rbf'}           0.966667   
3       linear  {'C': 30, 'kernel': 'linear'}           1.000000   
4          rbf     {'C': 50, 'kernel': 'rbf'}           1.000000   
5       linear  {'C': 50, 'kernel': 'linear'}           1.000000   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           1.000000           0.966667           0.966667                1.0   
1           1.000000           0.966667           0.966667                1.0   
2           1.000000           0.900000           0.933333                1.0   
3           1.000000           0.900000           0.900000                1.0   
4           0.966667           0.900000           0.933333                1.0   
5           1.000000           0.900000           0.933333                1.0   

   mean_test_score  std_test_score  rank_test_score  
0         0.980000        0.016330                1  
1         0.980000        0.016330                1  
2         0.960000        0.038873                4  
3         0.960000        0.048990                4  
4         0.960000        0.038873                4  
5         0.966667        0.042164                3

In [17]:
df[['param_C', 'param_kernel', 'mean_test_score']]

param_C param_kernel  mean_test_score
0        1          rbf         0.980000
1        1       linear         0.980000
2       30          rbf         0.960000
3       30       linear         0.960000
4       50          rbf         0.960000
5       50       linear         0.966667

In [18]:

clf.best_score_

np.float64(0.9800000000000001)

### **Use RandomizedSearchCV to reduce number of iterations and with random combination of parameters. This is useful when you have too many parameters to try and your training time is longer. It helps reduce the cost of computation**

In [22]:
from sklearn.model_selection import RandomizedSearchCV
rs= RandomizedSearchCV(svm.SVC(gamma= 'auto'),
                       {
                           'C': [1, 30, 50],
                           'kernel': ['rbf', 'linear']
                       },
                       cv= 5,
                       return_train_score=False,
                       n_iter=2)
rs.fit(iris.data, iris.target)
pd.DataFrame(rs.cv_results_)[['param_C', 'param_kernel', 'mean_test_score']]

param_C param_kernel  mean_test_score
0        1       linear         0.980000
1       50       linear         0.966667

### **How about different models with different hyperparameters?**

In [23]:
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

model_params= {
    'svm': {
        'model': svm.SVC(gamma= 'auto'),
        'params': {
            'C': [1, 40, 70],
            'kernel': ['rbf', 'linear']
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params': {
            'n_estimators': [1, 5, 10]
        }
    },

    'logistic_regression': {
        'model': LogisticRegression(solver= 'liblinear', multi_class= 'auto'),
        'params': {
            'C': [1, 5, 10]
        }
    }
}


In [33]:
scores=[]

for model_name, mp in model_params.items():
  clf= GridSearchCV(mp['model'], mp['params'], cv= 5, return_train_score= False)
  clf.fit(iris.data, iris.target)
  scores.append({
      'model': model_name,
      'best_score': clf.best_score_,
      'best_params': clf.best_params_
  })

df= pd.DataFrame(scores, columns= ['model', 'best_score',  'best_params'])
df

model  best_score                best_params
0                  svm    0.980000  {'C': 1, 'kernel': 'rbf'}
1        random_forest    0.966667        {'n_estimators': 5}
2  logistic_regression    0.966667                   {'C': 5}

#### **Based on above, I can conclude that SVM with C=1 and kernel='rbf' is the best model for solving my problem of iris flower classification**

In [34]:
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier


model_params= {
    'svm': {
        'model': svm.SVC(gamma= 'auto'),
        'params': {
            'C': [1, 40, 70],
            'kernel': ['rbf', 'linear']
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params': {
            'n_estimators': [1, 5, 10]
        }
    },

    'logistic_regression': {
        'model': LogisticRegression(solver= 'liblinear', multi_class= 'auto'),
        'params': {
            'C': [1, 5, 10]
        }
    },
    'gaussian_nb': {
        'model': GaussianNB(),
        'params': {}
    },

    'multinomial_nb': {
        'model': MultinomialNB(),
        'params': {}
    },
    'decision_tree': {
        'model': DecisionTreeClassifier(),
        'params': {
            'criterion': ['gini', 'entropy']
        }
    }
}



In [36]:
scores=[]

for model_name, mp in model_params.items():
  clf= GridSearchCV(mp['model'], mp['params'], cv= 5, return_train_score=False)
  clf.fit(iris.data, iris.target)
  scores.append({
      'model': model_name,
      'best_score': clf.best_score_,
      'best_params': clf.best_params_
  })

df= pd.DataFrame(scores, columns=['model', 'best_score', 'best_params'])
df

model  best_score                best_params
0                  svm    0.980000  {'C': 1, 'kernel': 'rbf'}
1        random_forest    0.966667       {'n_estimators': 10}
2  logistic_regression    0.966667                   {'C': 5}
3          gaussian_nb    0.953333                         {}
4       multinomial_nb    0.953333                         {}
5        decision_tree    0.960000      {'criterion': 'gini'}

**For me the winner is svm (C=1, kernel=linear) with 98% score.**